### ISSUE TITLE
Indexing mapped type with keyof type param

### ISSUE THREAD

**Issue by liam-goodacre-hpe** [Microsoft/TypeScript#16018]
**TypeScript Version:**  2.2.2

**Code**

[On playground](http://www.typescriptlang.org/play/index.html#src=type%20Eg%20=%20{%20foo:%20string,%20bar:%20number%20}%0D%0A%0D%0Atype%20VariantRec<T>%20=%20{[K%20in%20keyof%20T]:%20{%20type:%20K,%20value:%20T[K]%20}}%0D%0Atype%20Alpha1%20=%20VariantRec<Eg>[keyof%20Eg]%0D%0A%0D%0Atype%20VariantType<T>%20=%20VariantRec<T>[keyof%20T]%0D%0Atype%20Alpha2%20=%20VariantType<Eg>%0D%0A%0D%0Aconst%20assignable12:%20Alpha2%20=%20null%20as%20Alpha1%0D%0Aconst%20assignable21:%20Alpha1%20=%20null%20as%20Alpha2%0D%0A)

```ts
type Eg = { foo: string, bar: number }

type VariantRec<T> = {[K in keyof T]: { type: K, value: T[K] }}
type Alpha1 = VariantRec<Eg>[keyof Eg]

type VariantType<T> = VariantRec<T>[keyof T]
type Alpha2 = VariantType<Eg>

const assignable12: Alpha2 = null as Alpha1
const assignable21: Alpha1 = null as Alpha2
```

**Expected behavior:**

Types `Alpha1` and `Alpha2` should be the identical.

**Actual behavior:**

`Alpha2` incorrectly becomes: `{ type: 'foo' | 'bar', value: string | number }`
It should be: `{ type: 'foo', value: string } | { type: 'bar', value: number }`

**Comment by tycho01** [comment]
Interesting. I'd posted a similar issue at #16244.

**Comment by tycho01** [comment]
Hey, we found a solution for mine by separating the steps using [generic defaults](https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-307614501); still kinda feels like a bug, but I think a workaround like that may solve your case too! 😃

**Comment by liam-goodacre-hpe** [comment]
@tycho01 Awesome, thanks!  That appears to work 😸  Here's an updated [playground](http://www.typescriptlang.org/play/index.html#src=type%20Eg%20=%20{%20foo:%20string,%20bar:%20number%20}%0D%0A%0D%0Atype%20VariantRec<T>%20=%20{[K%20in%20keyof%20T]:%20{%20type:%20K,%20value:%20T[K]%20}}%0D%0Atype%20Alpha1%20=%20VariantRec<Eg>[keyof%20Eg]%0D%0A%0D%0Atype%20VariantType<T,%20Tmp%20extends%20VariantRec<T>%20=%20VariantRec<T>>%20=%20Tmp[keyof%20T]%0D%0Atype%20Alpha2%20=%20VariantType<Eg>%0D%0A%0D%0Aconst%20assignable12:%20Alpha2%20=%20null%20as%20Alpha1%0D%0Aconst%20assignable21:%20Alpha1%20=%20null%20as%20Alpha2%0D%0A).

**Comment by tycho01** [comment]
Coming over from #16244, I just ran into another one super similar to yours, though here it doesn't seem fixable by defaults:
```ts
type Inc = [1, 2, 3, 4, 5];
type ObjectHasKey<O extends {}, K extends string> = ({[K in keyof O]: '1' } & { [k: string]: '0' })[K];
type Length<R extends {}, I extends number = 0> = { 1: Length<R, Inc[I]>, 0: I }[ObjectHasKey<R, I>];
type TestLenA = Length<{ 0: 'a', 1: 'b' }>;
// ^ ok, 2
type Wrap<O> = Length<O>;
// type Wrap<O, Len extends Length<O> = Length<O>> = Len;
// ^ old work-around, same result here
type TestLenB = Wrap<{ 0: 'a', 1: 'b' }>;
// ^ 0 :(
```
With the `wrap` wording I'm a bit reminded of #10247, though that concerned wrapping generic functions, while this is wrapping a generic type. Unfortunately, while in that other case one could observe the degenerating generics in the transformed output function, over here the inner workings feel somewhat opaque...

**Comment by mhegazy** [comment]
seems like a duplicate of https://github.com/Microsoft/TypeScript/issues/15756

**Comment by tycho01** [comment]
Fixed by #18042.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
